### DISCUSSION TITLE
How to measure NER model's loss in a validation set?

### DISCUSSION THREAD

**Discussion by echatzikyriakidis** [explosion/spaCy#6574] [Help: Coding & Implementations] [answered]
How to measure **NER** model's loss in a validation set?

I have it from training set but I want also for validation set to plot train/val losses per epoch and inspect bias/variance tradeoff.

Can someone give an example?

Keras and Tensorflow usually has an extra parameter for val_X when fitting a model.

However, all the examples from spaCy use no validation set.

I know that the `train` CLI command supports it but I have implemented my own train function and I want to know how this can be done.

This is what I have so far and I want to introduct support for `val_X` argument like `train_X`:

```
def train_blank_ner_model(language_id, train_X, entity_types, epochs, random_state, dropout, minibatch_size, losses_display_frequency_in_epochs):
  fix_random_seed(random_state)

  nlp = spacy.blank(language_id)

  assert len(nlp.pipe_names) == 0, f"Pipeline of blank model '{language_id}' is not empty."

  ner = nlp.create_pipe('ner')

  nlp.add_pipe(ner)

  for entity_type in entity_types:
    ner.add_label(entity_type)

  optimizer = nlp.begin_training()

  for epoch in tqdm(range(1, epochs + 1)):
    train_X = shuffle(train_X, random_state=random_state)

    losses = {}

    batches = minibatch(train_X, size=compounding(*minibatch_size))

    for batch in tqdm(batches, leave=False):
      texts, annotations = zip(*batch)
  
      nlp.update(texts, annotations, sgd=optimizer, drop=dropout, losses=losses)

    if epoch % losses_display_frequency_in_epochs == 0:
      print(f"Epoch {epoch}, Loss: {losses['ner']}")

  print(f"Training completed with loss: {losses['ner']}")

  return nlp
```
Thank you!

**Comment by svlandeg** [comment, marked as answer]
The training CLI uses the [Scorer](https://spacy.io/api/scorer#_title) to obtain accuracy (and other) measures on a development set. You can access this functionality with [`nlp.evaluate`](https://spacy.io/api/language#evaluate). I personally think these values are more meaningful than some arbitrary loss value that depends on the exact implementation of the algorithm. Personally, I like to monitor a training run for the point in time when the training loss is still going down, but the performance on the dev set starts getting worse - this is where you get into overfitting and where you likely want to stop training further.

**Reply by echatzikyriakidis** [reply]
Hi @svlandeg!

You helped a lot!

The last hours I play with with `nlp.evaluate` to solve this. How do you think the new version of the code (see below)? Now, I know in each epoch the train loss + train/validation P/R/F-score metrics.

Now, I can do a lot of things. Plot train/val F-score per epochs. Check if the loss decreases while the validation F-score is not improving any more or even decreases (from what you suggested).

However, there is always a "But..."! It introduces some delay.

My dataset is 20000 examples and split ratio of validation_set is 15%

`nlp.evaluate` delays some time to complete for training set and validation set. 6 minutes in total. Assuming I want to run 20 epochs it is 2 hours extra wait time.

How can I make it faster? I know that `nlp.evaluate` uses internal nlp.pipe for multiprocessing but in Colab I get only 2 CPU cores and nlp.evaluate can't be very fast.

Any idea how to improve its performance?

```
def train_blank_ner_model(language_id, train_X, val_X, entity_types, epochs, random_state, dropout, minibatch_size, capture_losses_frequency_in_epochs):
  fix_random_seed(random_state)

  nlp = spacy.blank(language_id)

  assert len(nlp.pipe_names) == 0, f"Pipeline of blank model '{language_id}' is not empty."

  ner = nlp.create_pipe('ner')

  nlp.add_pipe(ner)

  for entity_type in entity_types:
    ner.add_label(entity_type)

  optimizer = nlp.begin_training()

  metrics = { 'epoch' : [], 'train_loss' : [], 'train_fscore' : [], 'validation_fscore' : [], 'train_precision' : [], 'validation_precision' : [], 'train_recall' : [], 'validation_recall' : [] }

  print("{:^30}{:^30}{:^30}{:^30}{:^30}{:^30}{:^30}{:^30}".format('epoch', 'train_loss', 'train_fscore', 'validation_fscore', 'train_precision', 'validation_precision', 'train_recall', 'validation_recall'))

  for epoch in tqdm(range(1, epochs + 1)):
    train_X = shuffle(train_X, random_state=random_state)

    train_losses = {}

    batches = minibatch(train_X, size=compounding(*minibatch_size))

    for batch in tqdm(batches, leave=False):
      texts, annotations = zip(*batch)
  
      nlp.update(texts, annotations, sgd=optimizer, drop=dropout, losses=train_losses)

    if epoch % capture_losses_frequency_in_epochs == 0:
      train_scores = nlp.evaluate(train_X).scores

      val_scores = nlp.evaluate(val_X).scores

      print("{:^30}{:^30.3f}{:^30.3f}{:^30.3f}{:^30.3f}{:^30.3f}{:^30.3f}{:^30.3f}".format(epoch, train_losses['ner'], train_scores['ents_f'], val_scores['ents_f'], train_scores['ents_p'], val_scores['ents_p'], train_scores['ents_r'], val_scores['ents_r']))

      metrics['epoch'].append(epoch)
      metrics['train_loss'].append(train_losses['ner'])
      metrics['train_precision'].append(train_scores['ents_p'])
      metrics['train_recall'].append(train_scores['ents_r'])
      metrics['train_fscore'].append(train_scores['ents_f'])
      metrics['validation_precision'].append(val_scores['ents_p'])
      metrics['validation_recall'].append(val_scores['ents_r'])
      metrics['validation_fscore'].append(val_scores['ents_f'])

  return { 'model': nlp, 'metrics': pd.DataFrame(metrics) }
```

**Comment by svlandeg** [comment]
Have you tried playing with the `batch_size` parameter of `nlp.evaluate`?

**Reply by echatzikyriakidis** [reply]
@svlandeg No, I do not not pass any specific value. How this can affect the time needed?

One solution could be: Maybe running `nlp.evaluate` only in validation set which will always is going to be a small subset. I still will be able to understand if the model overfits, from what you suggested: The point where train loss decreases but validation F-score starts to decrease.

What do you think?

**Reply by echatzikyriakidis** [reply]
Or maybe I can just live with this delay and accept it as is.

Can you please tell me, the train CLI command does output also train P/R/F-score metrics or only in development set?

**Reply by svlandeg** [reply]
It could be that processing larger batches would speed up your processing time. But it also depends on the size of your documents and the memory required, so I'd advice to experiment a bit.

You can definitely try a smaller validation set as well. It depends on the variability in your dataset, what a good (representative) number of documents would be.

**Reply by svlandeg** [reply]
The train command outputs the training loss, and all other metrics (accuracy etc) are measured on the dev set.

**Reply by echatzikyriakidis** [reply]
This is interesting. I will remove train metrics, except train loss which is already there from `nlp
update`. Maybe this ia yhe reason why train ClI does not calculates train metrics. If the train set is large there will be a large delay. But validation is controllable, you can always decrease it and make it any size. How do you feel on this?

**Reply by echatzikyriakidis** [reply]
Currently, statistically my 95% of docs are 3000 characters in average. There might be some other with very big length.

**Reply by echatzikyriakidis** [reply]
I see default batch size is 256. I think it is nice. In GPU mode you cannot increase it on whatever size you want. I think there are some limitations.

**Reply by echatzikyriakidis** [reply]
@svlandeg Thank you!

I haved decided to measure metrics only on validation set, like train CLI command does.

Also, currently testing different batch size (double in size) to check if affects time.

The validation set metrics delay only for 1m. I can live with it.

I will do monitoring of train loss and validation F-score for checking overfitting or optimization plateau.

Stay safe.

**Reply by svlandeg** [reply]
Happy to hear you've found a solution that works on your use-case!
